checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java

```diff
@@ -77,6 +79,37 @@ private void knownToBeLessThanLengthOf(
             knownToBeArrayLength((NumericalAdditionNode) node, array, in, store);
         } else if (node instanceof NumericalSubtractionNode) {
             knownToBeArrayLength((NumericalSubtractionNode) node, array, in, store);
+        } else if (node instanceof NumericalMultiplicationNode) {
+            Node right = ((NumericalMultiplicationNode) node).getRightOperand();
+            Node left = ((NumericalMultiplicationNode) node).getLeftOperand();
+            knownToBeArrayLengthMultiplication(right, left, array, in, store);
+            knownToBeArrayLengthMultiplication(left, right, array, in, store);
+        }
+    }
+
+    /**
+     * The positive node times node is known to be exactly the length of arrayExp. If the positive
+     * node is really positive, then node is less than or equal to the length of arrayExp. If
+     * positive is greater than 1, then node is less than the length of arrayExp.
```

Comment:
This was hard for me to interpret on first reading.  Can you put the identifiers within `{@code ...}` to help readers parse the text?